File: apps/android/domain/src/main/kotlin/com/sanvya/app/domain/insights/Dividends.kt
package com.sanvya.app.domain.insights

import com.sanvya.app.domain.ledger.RateLookup
import java.time.Instant
import java.time.LocalDate
import java.time.ZoneOffset
import kotlin.math.roundToLong

// Ported verbatim from apps/web/src/market/dividends.ts (89 lines) for the
// Insights feed's dividend_income/portfolio_projection cards (task #28).
// Neither this file's callers, nor apps/investments (P3.10/P3.15, task
// #26/#40), previously ported this math -- Investments' own spec lists a
// Dividend/Projection panel that was never actually built; this is the
// first mobile port of dividends.ts, done here because the Insights cards
// need it. Not golden-vector tested (dividends.ts has no vectors on web
// either -- see AUDIT_HISTORY's Insights entry).

/** A holding, reduced to the fields dividend matching needs. */
data class HoldingLite(val symbol: String, val exchange: String?, val quantity: Double, val currency: String)

/** One row from `market_dividends` (global, read-only). */
data class DivRow(val symbol: String, val exchange: String?, val exDate: String, val payDate: String?, val amount: Long, val currency: String)

/** One dividend payment estimated in the user's base currency (minor units). */
data class DivEvent(val date: String, val base: Long, val upcoming: Boolean)

private fun divKey(symbol: String, exchange: String?) = "${symbol.uppercase()}|${(exchange ?: "").uppercase()}"

/**
 * Estimate dividend income per ex-date in base currency: for each dividend
 * row, sum (amount-per-share x shares held) over matching holdings,
 * converted to base. Uses CURRENT quantity (historical share counts aren't
 * tracked -- a reasonable estimate, matches web exactly). Matches on
 * symbol+exchange, falling back to symbol only.
 */
fun computeDividendEvents(holdings: List<HoldingLite>, dividends: List<DivRow>, getRate: RateLookup, base: String): List<DivEvent> {
    val bySymEx = LinkedHashMap<String, MutableList<HoldingLite>>()
    val bySym = LinkedHashMap<String, MutableList<HoldingLite>>()
    for (h in holdings) {
        bySymEx.getOrPut(divKey(h.symbol, h.exchange)) { mutableListOf() }.add(h)
        bySym.getOrPut(h.symbol.uppercase()) { mutableListOf() }.add(h)
    }
    val today = LocalDate.now(ZoneOffset.UTC).toString()
    val events = mutableListOf<DivEvent>()
    for (d in dividends) {
        val matches = bySymEx[divKey(d.symbol, d.exchange)] ?: bySym[d.symbol.uppercase()] ?: emptyList()
        if (matches.isEmpty()) continue
        val shares = matches.sumOf { it.quantity }
        if (shares <= 0.0) continue
        val inCcy = d.amount * shares
        val rate = if (d.currency == base) 1.0 else getRate(d.currency, base)
        events.add(DivEvent(date = d.exDate, base = (inCcy * rate).roundToLong(), upcoming = d.exDate >= today))
    }
    return events.sortedBy { it.date }
}

enum class DividendPeriod { WEEK, MONTH, QUARTER, YEAR, ALL }

data class DividendBucket(val label: String, val key: String, val value: Long, val upcoming: Boolean)

private val MONTHS_SHORT = listOf("Jan", "Feb", "Mar", "Apr", "May", "Jun", "Jul", "Aug", "Sep", "Oct", "Nov", "Dec")

private fun isoWeek(d: LocalDate): String {
    val week = d.get(java.time.temporal.WeekFields.ISO.weekOfWeekBasedYear())
    val wby = d.get(java.time.temporal.WeekFields.ISO.weekBasedYear())
    return "$wby-W${week.toString().padStart(2, '0')}"
}

/** Group events into period buckets. Recent windows are capped; "all" spans
 * everything by year -- matches web's bucketize() exactly. */
fun bucketize(events: List<DivEvent>, period: DividendPeriod): List<DividendBucket> {
    val map = LinkedHashMap<String, DividendBucket>()
    fun put(k: String, label: String, v: Long, upcoming: Boolean) {
        val cur = map[k]
        map[k] = if (cur != null) cur.copy(value = cur.value + v, upcoming = cur.upcoming || upcoming) else DividendBucket(label, k, v, upcoming)
    }
    for (e in events) {
        val d = LocalDate.parse(e.date.take(10))
        when (period) {
            DividendPeriod.WEEK -> put(isoWeek(d), "${MONTHS_SHORT[d.monthValue - 1]} ${d.dayOfMonth}", e.base, e.upcoming)
            DividendPeriod.MONTH -> put("${d.year}-${d.monthValue.toString().padStart(2, '0')}", "${MONTHS_SHORT[d.monthValue - 1]} '${d.year.toString().takeLast(2)}", e.base, e.upcoming)
            DividendPeriod.QUARTER -> { val q = (d.monthValue - 1) / 3 + 1; put("${d.year}-Q$q", "Q$q '${d.year.toString().takeLast(2)}", e.base, e.upcoming) }
            else -> put(d.year.toString(), d.year.toString(), e.base, e.upcoming) // year & all -> by year
        }
    }
    val all = map.values.sortedBy { it.key }
    val cap = when (period) { DividendPeriod.WEEK -> 12; DividendPeriod.MONTH -> 12; DividendPeriod.QUARTER -> 8; DividendPeriod.YEAR -> 6; DividendPeriod.ALL -> 999 }
    return if (all.size > cap) all.takeLast(cap) else all
}

data class DividendSummary(val trailing12: Long, val upcoming12: Long, val total: Long)

/** Trailing-12-month realized income + projected next-12-month income (from
 * scheduled + trailing run-rate) -- matches web's dividendSummary() exactly. */
fun dividendSummary(events: List<DivEvent>): DividendSummary {
    val now = Instant.now().toEpochMilli()
    val yearMs = 365L * 86_400_000L
    var trailing12 = 0L; var upcoming12 = 0L; var total = 0L
    for (e in events) {
        total += e.base
        val t = LocalDate.parse(e.date.take(10)).atStartOfDay(ZoneOffset.UTC).toInstant().toEpochMilli()
        if (t <= now && t >= now - yearMs) trailing12 += e.base
        if (t > now && t <= now + yearMs) upcoming12 += e.base
    }
    return DividendSummary(trailing12, upcoming12, total)
}

File: apps/ios/Domain/Sources/Domain/Dividends.swift
import Foundation

// Ported verbatim from apps/web/src/market/dividends.ts (89 lines) for the
// Insights feed's dividend_income/portfolio_projection cards (task #28).
// Mirrors Android's domain/insights/Dividends.kt added the same session.
// Neither this file's callers, nor apps/investments (P3.10/P3.15, task
// #26/#40), previously ported this math -- Investments' own spec lists a
// Dividend/Projection panel that was never actually built; this is the
// first mobile port of dividends.ts, done here because the Insights cards
// need it. Not golden-vector tested (dividends.ts has no vectors on web
// either -- see AUDIT_HISTORY's Insights entry).

/// A holding, reduced to the fields dividend matching needs.
public struct HoldingLite: Sendable {
    public let symbol: String
    public let exchange: String?
    public let quantity: Double
    public let currency: String
    public init(symbol: String, exchange: String?, quantity: Double, currency: String) {
        self.symbol = symbol; self.exchange = exchange; self.quantity = quantity; self.currency = currency
    }
}

/// One row from `market_dividends` (global, read-only).
public struct DivRow: Sendable {
    public let symbol: String
    public let exchange: String?
    public let exDate: String
    public let payDate: String?
    public let amount: Int64
    public let currency: String
    public init(symbol: String, exchange: String?, exDate: String, payDate: String?, amount: Int64, currency: String) {
        self.symbol = symbol; self.exchange = exchange; self.exDate = exDate; self.payDate = payDate; self.amount = amount; self.currency = currency
    }
}

/// One dividend payment estimated in the user's base currency (minor units).
public struct DivEvent: Sendable {
    public let date: String
    public let base: Int64
    public let upcoming: Bool
}

private func divKey(_ symbol: String, _ exchange: String?) -> String {
    "\(symbol.uppercased())|\((exchange ?? "").uppercased())"
}

/// Estimate dividend income per ex-date in base currency: for each dividend
/// row, sum (amount-per-share x shares held) over matching holdings,
/// converted to base. Uses CURRENT quantity (historical share counts
/// aren't tracked -- a reasonable estimate, matches web exactly). Matches
/// on symbol+exchange, falling back to symbol only.
public func computeDividendEvents(_ holdings: [HoldingLite], _ dividends: [DivRow], _ getRate: RateLookup, _ base: String) -> [DivEvent] {
    var bySymEx: [String: [HoldingLite]] = [:]
    var bySym: [String: [HoldingLite]] = [:]
    for h in holdings {
        bySymEx[divKey(h.symbol, h.exchange), default: []].append(h)
        bySym[h.symbol.uppercased(), default: []].append(h)
    }
    let today = String(ISO8601DateFormatter().string(from: Date()).prefix(10))
    var events: [DivEvent] = []
    for d in dividends {
        let matches = bySymEx[divKey(d.symbol, d.exchange)] ?? bySym[d.symbol.uppercased()] ?? []
        if matches.isEmpty { continue }
        let shares = matches.reduce(0.0) { $0 + $1.quantity }
        if shares <= 0 { continue }
        let inCcy = Double(d.amount) * shares
        let rate = d.currency == base ? 1.0 : getRate(d.currency, base)
        events.append(DivEvent(date: d.exDate, base: Int64((inCcy * rate).rounded()), upcoming: d.exDate >= today))
    }
    return events.sorted { $0.date < $1.date }
}

public enum DividendPeriod: Sendable { case week, month, quarter, year, all }

public struct DividendBucket: Sendable {
    public let label: String
    public let key: String
    public let value: Int64
    public let upcoming: Bool
}

private let MONTHS_SHORT = ["Jan", "Feb", "Mar", "Apr", "May", "Jun", "Jul", "Aug", "Sep", "Oct", "Nov", "Dec"]

private func isoWeek(_ d: Date) -> String {
    var cal = Calendar(identifier: .iso8601)
    cal.timeZone = TimeZone(identifier: "UTC")!
    let week = cal.component(.weekOfYear, from: d)
    let year = cal.component(.yearForWeekOfYear, from: d)
    return "\(year)-W\(String(format: "%02d", week))"
}

/// Group events into period buckets. Recent windows are capped; "all"
/// spans everything by year -- matches web's bucketize() exactly.
public func bucketize(_ events: [DivEvent], _ period: DividendPeriod) -> [DividendBucket] {
    var map: [String: DividendBucket] = [:]
    var order: [String] = []
    func put(_ k: String, _ label: String, _ v: Int64, _ upcoming: Bool) {
        if let cur = map[k] {
            map[k] = DividendBucket(label: cur.label, key: cur.key, value: cur.value + v, upcoming: cur.upcoming || upcoming)
        } else {
            map[k] = DividendBucket(label: label, key: k, value: v, upcoming: upcoming)
            order.append(k)
        }
    }
    var cal = Calendar(identifier: .gregorian)
    cal.timeZone = TimeZone(identifier: "UTC")!
    for e in events {
        guard let d = ISO8601DateFormatter().date(from: e.date + "T00:00:00Z") else { continue }
        let year = cal.component(.year, from: d); let month = cal.component(.month, from: d)
        switch period {
        case .week: put(isoWeek(d), "\(MONTHS_SHORT[month - 1]) \(cal.component(.day, from: d))", e.base, e.upcoming)
        case .month: put("\(year)-\(String(format: "%02d", month))", "\(MONTHS_SHORT[month - 1]) '\(String(year).suffix(2))", e.base, e.upcoming)
        case .quarter:
            let q = (month - 1) / 3 + 1
            put("\(year)-Q\(q)", "Q\(q) '\(String(year).suffix(2))", e.base, e.upcoming)
        case .year, .all: put(String(year), String(year), e.base, e.upcoming) // year & all -> by year
        }
    }
    let all = order.sorted().compactMap { map[$0] }
    let cap: Int
    switch period { case .week: cap = 12; case .month: cap = 12; case .quarter: cap = 8; case .year: cap = 6; case .all: cap = 999 }
    return all.count > cap ? Array(all.suffix(cap)) : all
}

public struct DividendSummary: Sendable {
    public let trailing12: Int64
    public let upcoming12: Int64
    public let total: Int64
}

/// Trailing-12-month realized income + projected next-12-month income (from
/// scheduled + trailing run-rate) -- matches web's dividendSummary() exactly.
public func dividendSummary(_ events: [DivEvent]) -> DividendSummary {
    let now = Date().timeIntervalSince1970 * 1000
    let yearMs: Double = 365 * 86_400_000
    var trailing12: Int64 = 0; var upcoming12: Int64 = 0; var total: Int64 = 0
    for e in events {
        total += e.base
        guard let d = ISO8601DateFormatter().date(from: e.date + "T00:00:00Z") else { continue }
        let t = d.timeIntervalSince1970 * 1000
        if t <= now && t >= now - yearMs { trailing12 += e.base }
        if t > now && t <= now + yearMs { upcoming12 += e.base }
    }
    return DividendSummary(trailing12: trailing12, upcoming12: upcoming12, total: total)
}
